### ISSUE TITLE
[Bug]: GPU plugin miscompiles Qwen3-VL vision merger when the token dimension is dynamic

### ISSUE THREAD

**Issue by cyberluke** [openvinotoolkit/openvino#37573]
### OpenVINO Version

master @ `7c4d2733dc0521f0f1203f3664df305905a80095` (built from source). Also
reproduced on the nightly wheel `2026.4.0.dev20260821` (build 22849).

### Operating System

Windows System (Windows 11 Pro Insider Preview, build 26300)

### Device used for inference

GPU

### Framework

PyTorch

### Model used

[OpenVINO/Qwen3-VL-8B-Instruct-int4-ov](https://huggingface.co/OpenVINO/Qwen3-VL-8B-Instruct-int4-ov)
- specifically `openvino_vision_embeddings_merger_model.xml`

### Issue description

The GPU plugin produces numerically wrong results for the Qwen3-VL vision
merger graph **when the vision-token dimension is dynamic**, which is how the
model is exported. Reshaping the identical IR to a static token count restores
CPU parity.

Same IR, same inputs, CPU as oracle:

| configuration | cosine vs CPU | mean abs | verdict |
|---|---|---|---|
| GPU, dynamic token dim (as exported) | **0.146181** | 1.5666 | wrong |
| GPU, same IR reshaped to static | 0.998810 | 0.1287 | correct |
| CPU reference | 1.0 | 0.1292 | - |

There is **no exception and no warning** - the graph runs and returns
plausible-looking tensors whose magnitudes are roughly 12x inflated. In a VLM
pipeline this surfaces as a model that answers fluently while ignoring the
image, which is considerably harder to notice than a crash.

#### Localisation

Bisecting the merger graph (1586 instrumentable ops), then verifying by
comparing every direct input of the divergent node:

```
first divergent op:
  __module.blocks.0.attn/aten::scaled_dot_product_attention/ScaledDotProductAttention
  output (16, 2880, 72) f32
  q    cosine 0.999998
  k    cosine 0.999994
  v    cosine 0.999980
  mask cosine 1.000000 (exact)
  out  cosine 0.052697   max_abs 6.18
```

Correct inputs in, wrong output out, GPU only.

What this is **not**:

- **Not precision.** `INFERENCE_PRECISION_HINT=f32` was verified applied
  (`get_property` returns `float32`, all 25 runtime ops report `f32`) and the
  result is unchanged (`0.052697` -> `0.052746`). No NaN/Inf.
- **Not the SDPA primitive.** Forcing the attention unfused
  (`Gemm -> SoftMax -> Gemm`, confirmed via `get_runtime_model()` that no
  `scaled_dot_product_attention` primitive remains) still reproduces. A
  standalone SDPA at the same shapes - static or dynamic, `head_dim` 64/72/80/128,
  seq 256/2880 - is correct on GPU (cosine > 0.99999).
- **Not the export.** All three exported graphs execute correctly on CPU, and
  the language model returns finite logits when fed hand-built inputs.

It is graph-context dependent. Pruning the merger to the 79-op subgraph that
feeds that SDPA still reproduces (`cosine 0.052697`). Replacing any upstream
input with a **static** `Parameter` - either above or below the `FullyConnected`
- makes the divergence disappear, which is what pointed at the dynamic
dimension rather than at any single operation.

Note that a decomposition written by hand is re-fused by the plugin
(`get_runtime_model()` still shows `scaled_dot_product_attention`), so a
decomposition test must be verified rather than assumed.

### Step-by-step reproduction

Self-contained reproducer (random inputs, public model, no image data needed):

**https://gist.github.com/cyberluke/65d3206c0c2bad2f419bfb1c81121187**

```bash
pip install openvino numpy huggingface_hub
python repro_gpu_dynamic_merger.py
```

It downloads `openvino_vision_embeddings_merger_model.{xml,bin}` from the public
repository, runs the merger on CPU and on GPU twice (as-exported dynamic shapes,
then the same IR reshaped static), and prints cosine similarity against the CPU
result.

Expected on an affected setup: dynamic is wrong, static is correct.

### Relevant log output

```shell
openvino : 2026.4.0-1-7c4d2733dc0
devices  : ['CPU', 'GPU.0', 'GPU.1']

CPU reference: shape=(720, 4096) mean|x|=0.1292
GPU, dynamic (as exported)  cosine_vs_CPU= 0.146181  max_abs_diff= 12.1952  mean|x|=1.5666  WRONG
GPU, static reshape         cosine_vs_CPU= 0.998810  max_abs_diff=  0.5599  mean|x|=0.1287  OK
```

Environment:

```shell
CPU    : Intel(R) Core(TM) Ultra 9 285K
GPU.0  : Intel(R) Graphics (iGPU)
driver : 32.0.101.8974 (2026-08-11)
OS     : Windows 11 Pro Insider Preview, build 26300
```

`GPU.1` is an NVIDIA card present in the system but not used; the defect is on
`GPU.0`.

### AI usage disclosure

```
AI assistance used: yes

If yes: This defect was found and localised with Kelvin Clyne, an agentic
        coding assistant, while qualifying Qwen3-VL-8B on an Intel Xe iGPU.
        AI was used to bisect the graph, verify the divergent node's inputs,
        delta-reduce the reproducer, and draft this report.

Human validation performed: OpenVINO built from source (master 7c4d2733,
        MSVC 14.44 / Ninja) and cross-checked against the 2026.4.0.dev20260821
        nightly wheel; every number above is a first-hand measurement on the
        hardware listed, reproduced from a clean process; the linked reproducer
        was executed end-to-end before publication.
```

### Issue submission checklist

- [x] I'm reporting an issue. It's not a question.
- [x] I checked the problem with the documentation, FAQ, open issues, Stack Overflow, etc., and have not found a solution.
- [x] There is reproducer code and related data files such as images, videos, models, etc.

**Comment by Zulkifli-Intel** [comment]
Hi @cyberluke 
I tested on my side and received the same result, the static GPU output is very close to the CPU output, while the dynamic GPU output is dramatically different, and GPU execution doesn't crash or report an error. I'm checking this internally with the dev team for further investigation.

**Comment by diego-villalobos** [comment]
Ref. 194066

**Comment by alvoron** [comment]
@cyberluke thank you for raising the issue!
I was able to reproduce the issue. Also I checked that the issue is already fixed in master by https://github.com/openvinotoolkit/openvino/pull/37712
Could you please double check using the nightly wheel or the latest master?
I closed the ticket. Feel free to reopen it if you reproduce the issue on the latest code base.

**Comment by cyberluke** [comment]
Thanks for confirming that #37573 is fixed by #37712.

I would like to clarify the provenance and attribution here, because the public timeline is a bit surprising.

I reported #37573 on August 23 with a standalone reproducer, CPU/GPU differential measurements, graph bisection to the first divergent SDPA node, direct-input verification, and elimination of precision, export, and standalone-SDPA explanations. Intel subsequently confirmed the reproduction and said the issue was being investigated internally.

The first public PR I can find for CVS-192936 is #37645 from August 26, followed by the actual fix in #37712 from August 28, both after my report. #37712 fixes the same Qwen3-VL GPU wrong-result path, but currently does not reference #37573 or the external report.

I understand that CVS-192936 may have existed internally before my report. If so, could you clarify when that issue was originally filed?

Either way, since #37573 independently reproduced and substantially localized the defect before the public fix existed, I would appreciate the external report being linked/credited from the fixing PR, for example as:

`Reported independently by @cyberluke in #37573`

or an equivalent reference.

I'm not asking for authorship of Intel's implementation. I am asking that the public provenance of the bug report and investigation is preserved.
---

I also want to add one broader point.

This is not how healthy open-source collaboration should work.

I am a CTO, and I invested a significant amount of my own engineering time, compute resources, hardware capacity, and financial resources into reproducing, isolating, validating, and documenting this defect properly. This was not a casual one-line bug report. The investigation included building OpenVINO from source, validating against nightly builds, constructing a standalone reproducer, performing graph-level bisection, numerical comparison, and narrowing the failure down far enough to give the GPU team a concrete and reproducible starting point.

When an external contributor puts that level of effort into improving an open-source project, and the resulting fix lands shortly afterwards without even linking the original report, it creates the impression that the contributor's work has simply disappeared into an internal engineering process.

I am not asking to take credit for code I did not write. I am asking for proper attribution for the work I did perform and for the public development history to accurately reflect how the issue was discovered, reproduced, and investigated.

Open source depends on contributors being able to trust that substantial external work will remain visible and attributable. Without that, there is very little incentive for experienced engineers to spend serious time and money investigating difficult defects upstream instead of simply maintaining private patches.

---

Also you send comment and 2 minutes later you close the issue with text: "I closed the ticket. Feel free to reopen it if you reproduce the issue on the latest code base." - but I have no permission to reopen the issue.

**Comment by cyberluke** [comment]
@alvoron

**Comment by alvoron** [comment]
@cyberluke 
thank you for the rising this and thank you for the work itself.
You are right, and the omission was ours. I have added the note `Reported independently by @cyberluke in #37573` to the PR, so the connection between your report and the fix is now part of the public history. I am sorry if was not there from the start.

For the completeness on status: the reported defect is fixed and first appears in the `2026.5.0.dev20260907` nightly; the decomposed-attention case you spotted is fixed in `2026.5.0.dev20260911` and later.

Thank you again for taking the time, the effort to investigate this issue. I'll keep the ticket open until you close it be yourself (as soon as the fix is confirmed on your side).

**Comment by cyberluke** [comment]
Perfect, thank you, I will retest.

**Comment by cyberluke** [comment]
Verified: the GPU dynamic-merger defect is resolved on the rebuilt master.
